- Study the correlation between features and find the features subset with the highest correlation with the target column, and try to explain from the business point of view why they have such a big correlation.
- Please comment on your code.

In [2]:
# import library
import numpy as np
import pandas as pd
# from sklearn import sklearn
import sklearn as sk

# library for corelation show
import seaborn as sns
# library for training
from sklearn.model_selection import train_test_split 
from sklearn.metrics import mean_squared_error, mean_absolute_error 
from sklearn.linear_model import LinearRegression


In [3]:
# import our data set
df = pd.read_csv('HousingData.csv')
df.head(10)

CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575   65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421   78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185   61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998   45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18   0.0  0.458  7.147   54.2  6.0622    3  222     18.7   
5  0.02985   0.0   2.18   0.0  0.458  6.430   58.7  6.0622    3  222     18.7   
6  0.08829  12.5   7.87   NaN  0.524  6.012   66.6  5.5605    5  311     15.2   
7  0.14455  12.5   7.87   0.0  0.524  6.172   96.1  5.9505    5  311     15.2   
8  0.21124  12.5   7.87   0.0  0.524  5.631  100.0  6.0821    5  311     15.2   
9  0.17004  12.5   7.87   NaN  0.524  6.004   85.9  6.5921    5  311     15.2   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90    NaN  36.2  
5  394.12   5.21  28.7  
6  395.60  12.43  22.9  
7  396.90  19.15  27.1  
8  386.63  29.93  16.5  
9  386.71  17.10  18.9

- Create a second set of data with the columns that have an absolute correlation between 0.5 and 0.8 with the target column.

In [4]:
# correlation describes the relationship between variables
# Pearson Correlation by default
df_coreleation = df.corr() # de făcut de la 0.5 - 0.8

df_coreleation

CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
CRIM     1.000000 -0.191178  0.401863 -0.054355  0.417130 -0.219150  0.354342   
ZN      -0.191178  1.000000 -0.531871 -0.037229 -0.513704  0.320800 -0.563801   
INDUS    0.401863 -0.531871  1.000000  0.059859  0.764866 -0.390234  0.638431   
CHAS    -0.054355 -0.037229  0.059859  1.000000  0.075097  0.104885  0.078831   
NOX      0.417130 -0.513704  0.764866  0.075097  1.000000 -0.302188  0.731548   
RM      -0.219150  0.320800 -0.390234  0.104885 -0.302188  1.000000 -0.247337   
AGE      0.354342 -0.563801  0.638431  0.078831  0.731548 -0.247337  1.000000   
DIS     -0.374166  0.656739 -0.711709 -0.093971 -0.769230  0.205246 -0.744844   
RAD      0.624765 -0.310919  0.604533  0.001468  0.611441 -0.209847  0.458349   
TAX      0.580595 -0.312371  0.731055 -0.032304  0.668023 -0.292048  0.509114   
PTRATIO  0.281110 -0.414046  0.390954 -0.111304  0.188933 -0.355501  0.269226   
B       -0.381411  0.171303 -0.360532  0.051264 -0.380051  0.128069 -0.275303   
LSTAT    0.444943 -0.414193  0.590690 -0.047424  0.582641 -0.614339  0.602891   
MEDV    -0.391363  0.373136 -0.481772  0.181391 -0.427321  0.695360 -0.394656   

              DIS       RAD       TAX   PTRATIO         B     LSTAT      MEDV  
CRIM    -0.374166  0.624765  0.580595  0.281110 -0.381411  0.444943 -0.391363  
ZN       0.656739 -0.310919 -0.312371 -0.414046  0.171303 -0.414193  0.373136  
INDUS   -0.711709  0.604533  0.731055  0.390954 -0.360532  0.590690 -0.481772  
CHAS    -0.093971  0.001468 -0.032304 -0.111304  0.051264 -0.047424  0.181391  
NOX     -0.769230  0.611441  0.668023  0.188933 -0.380051  0.582641 -0.427321  
RM       0.205246 -0.209847 -0.292048 -0.355501  0.128069 -0.614339  0.695360  
AGE     -0.744844  0.458349  0.509114  0.269226 -0.275303  0.602891 -0.394656  
DIS      1.000000 -0.494588 -0.534432 -0.232471  0.291512 -0.493328  0.249929  
RAD     -0.494588  1.000000  0.910228  0.464741 -0.444413  0.479541 -0.381626  
TAX     -0.534432  0.910228  1.000000  0.460853 -0.441808  0.536110 -0.468536  
PTRATIO -0.232471  0.464741  0.460853  1.000000 -0.177383  0.375966 -0.507787  
B        0.291512 -0.444413 -0.441808 -0.177383  1.000000 -0.369889  0.333461  
LSTAT   -0.493328  0.479541  0.536110  0.375966 -0.369889  1.000000 -0.735822  
MEDV     0.249929 -0.381626 -0.468536 -0.507787  0.333461 -0.735822  1.000000

In [5]:
# select the values that are bigger than 0.5
relevant_features_min = df_coreleation[df_coreleation > 0.5]  
# relevant_features_max = cor_target[cor_target < 0.8]

# relevant_features = df[(df['MEDV'] < 0.8) & (df['MEDV'] > 0.5)]
# relevant_features
print("Colomns > 0.5")
print(relevant_features_min)


Colomns > 0.5
             CRIM        ZN     INDUS  CHAS       NOX       RM       AGE  \
CRIM     1.000000       NaN       NaN   NaN       NaN      NaN       NaN   
ZN            NaN  1.000000       NaN   NaN       NaN      NaN       NaN   
INDUS         NaN       NaN  1.000000   NaN  0.764866      NaN  0.638431   
CHAS          NaN       NaN       NaN   1.0       NaN      NaN       NaN   
NOX           NaN       NaN  0.764866   NaN  1.000000      NaN  0.731548   
RM            NaN       NaN       NaN   NaN       NaN  1.00000       NaN   
AGE           NaN       NaN  0.638431   NaN  0.731548      NaN  1.000000   
DIS           NaN  0.656739       NaN   NaN       NaN      NaN       NaN   
RAD      0.624765       NaN  0.604533   NaN  0.611441      NaN       NaN   
TAX      0.580595       NaN  0.731055   NaN  0.668023      NaN  0.509114   
PTRATIO       NaN       NaN       NaN   NaN       NaN      NaN       NaN   
B             NaN       NaN       NaN   NaN       NaN      NaN       NaN  

In [21]:
cor = df.corr()
cor_target = abs(cor["MEDV"])

# sns.heatmap(cor, annot = True, cmap = plt.cm.Reds)
# plt.show()

relevant_features = cor_target[cor_target > 0.5]

# relevant_features
print("Colomns > 0.5")
print(relevant_features)


Colomns > 0.5
RM         0.695360
PTRATIO    0.507787
LSTAT      0.735822
MEDV       1.000000
Name: MEDV, dtype: float64


In [7]:
# de facut linear regresion peste toate coloanele - adici de seleecat y - medv, iar x sa fie restul coloanelor
# dupa care de caluclat eorare
# de vazut care coloane sunt intre intervalurile dorite adica 0.5 - 0.8 si de separat aprte si de facat linaear regresion
# dupa care de facut eroare
# la final de comparat ce si cum
# si dupa de facut concluzz

In [8]:
print(df[["LSTAT","PTRATIO"]].corr())
print(df[["RM","LSTAT"]].corr())

            LSTAT   PTRATIO
LSTAT    1.000000  0.375966
PTRATIO  0.375966  1.000000
             RM     LSTAT
RM     1.000000 -0.614339
LSTAT -0.614339  1.000000


- Split the data into 2 sub-sets using the train_test_split function from sklearn.

In [13]:
# exclude the Nan Values
df_new = df[np.isfinite(df).all(1)]

# create two subsets from all data sets
X = df_new[['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT']]
y = df_new[['MEDV']]

- Train a sklearn Linear Regression model on the data provided to you.

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3)

# Create an instance of the LinearRegression class from sklearn. 
lr = LinearRegression() 
    
# Call the fit function, which applies the Linear Regression model # onto the data. 
lr.fit(X_train, y_train) 
    
# Define a variable that stores the predictions of our model. 
y_pred = lr.predict(X_test) 
    
# Print the mean squared error and mean absolute error to get an idea of 
# how well the model has done. 
mean_squared_error(y_pred, y_test), mean_absolute_error( y_pred, y_test) 


(32.192322378935785, 3.499648482184587)

- Train a from-scratch implementation of Linear Regression on the train sub-set.

In [15]:
# SEED = 42
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

regressor = LinearRegression()

regressor.fit(X_train, y_train)

print(regressor.intercept_)

print(regressor.coef_)

# care sunt sub seturile pe care trebuie sa fac chestia data

[32.99322336]
[[-1.12990963e-01  4.49594371e-02  5.75447722e-02  1.18099414e+00
  -1.72522169e+01  4.27138350e+00 -1.99977624e-02 -1.40633199e+00
   2.77671705e-01 -1.68192666e-02 -8.96637094e-01  9.07231436e-03
  -3.64600546e-01]]


- Test the models on the test sets from the initial set of data, for error metrics use the models score function for the sklearn model.

In [16]:
# min value error - error - absolte, discret  - categ
from sklearn.metrics import mean_absolute_error,r2_score,mean_squared_error
def run_experiment(model):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    print("R^2 : ", r2_score(y_test, y_pred))
    print("MAE :", mean_absolute_error(y_test,y_pred))
    print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred)))

from sklearn.linear_model import LinearRegression
model = LinearRegression()
run_experiment(model)

# este corect?

R^2 :  0.6905175764205997
MAE : 3.4558210072479856
RMSE: 5.3731528852484205


- Split the data with the selected columns into 2 sub-sets using the train_test_split function from sklearn.

In [17]:
X = df_new[['RM', 'PTRATIO', 'LSTAT']] # de adaugat coloane 
y = df_new['MEDV']

- Train a sklearn Linear Regression model on the data with selected columns (train subset).

In [18]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3)

# Create an instance of the LinearRegression class from sklearn. 
lr = LinearRegression() 
    
# Call the fit function, which applies the Linear Regression model # onto the data. 
lr.fit(X_train, y_train) 
    
# Define a variable that stores the predictions of our model. 
y_pred = lr.predict(X_test) 
    
# Print the mean squared error and mean absolute error to get an idea of 
# how well the model has done. 
mean_squared_error(y_pred, y_test), mean_absolute_error(y_pred, y_test) 

# e corect aici?

(22.1172730105175, 3.4665169252620185)

- Train a from-scratch implementation of Linear Regression on the train sub-set.

In [19]:
# SEED = 42
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

regressor = LinearRegression()

regressor.fit(X_train, y_train)

print(regressor.intercept_)

print(regressor.coef_)

# care sunt sub seturile pe care trebuie sa fac chestia data

20.68817869575898
[ 4.40775241 -1.04711809 -0.53902903]


- Test the models on the test sets from the initial set of data, for error metrics use the models score function for the sklearn model.

In [20]:
# min value error - error - absolte, discret  - categ
from sklearn.metrics import mean_absolute_error,r2_score,mean_squared_error
def run_experiment(model):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    print("R^2 : ", r2_score(y_test, y_pred))
    print("MAE :", mean_absolute_error(y_test,y_pred))
    print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred)))

from sklearn.linear_model import LinearRegression
model = LinearRegression()
run_experiment(model)

# este corect?

R^2 :  0.6529769119096893
MAE : 3.5931643420930692
RMSE: 5.689713368772232


- Please try to interpret the results that you are getting by comparing the error of the models that you created.

### Result on Entire Dataframe

__Lineare Regresion:__ (32.192322378935785, 3.499648482184587)


__SEED:__ [32.99322336]

    [[-1.12990963e-01  4.49594371e-02  5.75447722e-02  1.18099414e+00 -1.72522169e+01  4.27138350e+00 -1.99977624e-02 -1.40633199e+00 2.77671705e-01 -1.68192666e-02 -8.96637094e-01  9.07231436e-03 -3.64600546e-01]]
  
__Error:__ 
    
    R^2 :  0.6905175764205997
    
    MAE : 3.4558210072479856
    
    RMSE: 5.3731528852484205

### Result on Target Columns

__Lineare Regresion:__ (22.1172730105175, 3.4665169252620185)


__SEED:__ 20.68817869575898

    [ 4.40775241 -1.04711809 -0.53902903]

__Error:__ 
    
    R^2 :  0.6529769119096893
    
    MAE : 3.5931643420930692
    
    RMSE: 5.689713368772232

### Observation
Dacă să luăm și să analizam ambele rezultatele la sigur putem observa diferențe, prima difrență pe care o avet este anume la calcularea linear regresion vedem că pentru întreg data set valorile LR sunt mai mare în comparație cu LR care a fost aplicată numai pe coloanele selectate.

Când facem train putem observă că iarăși valorile pentru întreg data set sunt mai mari.

Iar dacă să vorbim despre erori, eroare pentru întreg data set este mare mare în comparație cu eroare care este doar pentru coloanele noastre țintă